src/units/json.rs

```diff
@@ -191,54 +209,224 @@ impl TryFrom<String> for SourceUri {
             }
-            JsonReader::Http(ref mut inner) => {
-                return inner.read(buf)
+            SourceUri::File { ref path, ref mut last_modified } => {
+                Self::open_file(path, last_modified, component).await
             }
-            JsonReader::HttpRequest(ref mut inner) => {
-                match inner.take() {
-                    Some(inner) => {
-                        inner.send().map_err(|err| {
-                            io::Error::new(
-                                io::ErrorKind::Other,
-                                err
-                            )
-                        })?
-                    }
-                    None => {
-                        return Err(io::Error::new(
-                            io::ErrorKind::Other,
-                            "already failed to send request"
-                        ))
-                    }
+        }
+    }
+
+    async fn open_http(
+        uri: &Url, 
+        last_modified: &mut Option<DateTime<Utc>>,
+        etag: &mut Option<Bytes>,
+        component: &Component,
+    ) -> Result<Option<Self>, Failed> {
+        // Create and send the request.
+        let mut request = component.http_client().get(uri.clone());
+        if let Some(etag) = etag.as_ref() {
+            request = request.header(
+                header::IF_NONE_MATCH, etag.as_ref()
+            );
+        }
+        if let Some(ts) = last_modified {
+            request = request.header(
+                header::IF_MODIFIED_SINCE, format_http_date(*ts)
+            );
+        }
+        let response = request.send().await.map_err(|err| {
+            warn!(
+                "Unit {}: HTTP request failed: {}",
+                component.name(), err
+            );
+            Failed
+        })?;
+
+        // Return early if we anything other than a 200 OK
+        if response.status() == StatusCode::NOT_MODIFIED {
+            return Ok(None)
+        }
+        else if response.status() != StatusCode::OK {
+            warn!(
+                "Unit {}: HTTP request return status {}",
+                component.name(), response.status()
+            );
+            return Err(Failed)
+        }
+
+        // Update Etag and Last-Modified.
+        *etag = Self::parse_etag(&response);
+        *last_modified = Self::parse_last_modified(&response);
+
+        // And we are good to go!
+        Ok(Some(Self::new(ReaderSource::Http(response))))
+    }
+
+    async fn open_file(
+        path: &ConfigPath,
+        last_modified: &mut Option<SystemTime>,
+        component: &Component,
+    ) -> Result<Option<Self>, Failed> {
+        if let Ok(modified) = metadata(path).and_then(|meta| meta.modified()) {
+            if let Some(last_modified) = last_modified {
+                if *last_modified >= modified {
+                    return Ok(None)
+                }
+            }
+            *last_modified = Some(modified)
```

Comment:
Should you maybe wait with updating `last_modified` until after a successful read from the source?

Reply:
Yes, that seems like a better strategy.